frontend/src/test/scala/bloop/BuildLoaderSpec.scala

```diff
@@ -7,17 +7,129 @@ import bloop.engine.{Build, BuildLoader, State}
     }
   }
 
+  testLoad("reload if settings are added") { (testBuild, logger) =>
+    val settings = WorkspaceSettings("4.2.0", List(BuildInfo.scalaVersion))
+    testBuild.state.build.checkForChange(Some(settings), logger).map {
+      case Build.ReturnPreviousState =>
+        sys.error(s"Expected return updated state, got previous state")
+      case action: Build.UpdateState =>
+        assert(action.createdOrModified.size == 0)
+        assert(action.deleted.size == 0)
+        assert(action.invalidated.size == 4)
+        assert(action.settingsForReload == Some(settings))
+    }
+  }
+
+  val sameSettings = WorkspaceSettings("4.2.0", List(BuildInfo.scalaVersion))
+  testLoad("do not reload if same settings are added", Some(sameSettings)) { (testBuild, logger) =>
+    testBuild.state.build.checkForChange(Some(sameSettings), logger).map {
+      case Build.ReturnPreviousState => ()
+      case action: Build.UpdateState =>
+        pprint.log(action.invalidated)
+        sys.error(s"Expected return previous state, got updated state")
+    }
+  }
+
+  testLoad("reload if new settings are added", Some(sameSettings)) { (testBuild, logger) =>
+    val newSettings = WorkspaceSettings("4.1.11", List(BuildInfo.scalaVersion))
+    testBuild.state.build.checkForChange(Some(newSettings), logger).map {
+      case Build.ReturnPreviousState =>
+        sys.error(s"Expected return updated state, got previous state")
+      case action: Build.UpdateState =>
+        assert(action.createdOrModified.size == 0)
+        assert(action.deleted.size == 0)
+        assert(action.invalidated.size == 4)
+        assert(action.settingsForReload == Some(newSettings))
+    }
+  }
+
+  private def changeHashOfRandomFiles(build: TestBuild, n: Int): Unit = {
+    val randomFiles = scala.util.Random.shuffle(configurationFiles(build)).take(n)
+    randomFiles.foreach { f =>
+      // Add whitespace at the end to modify hash
+      val bytes = Files.readAllBytes(f.underlying)
+      val contents = new String(bytes, StandardCharsets.UTF_8) + " "
+      Files.write(f.underlying, contents.getBytes(StandardCharsets.UTF_8))
+    }
+  }
+
+  testLoad("reload if two file contents changed in build with previous", Some(sameSettings)) {
+    (testBuild, logger) =>
+      changeHashOfRandomFiles(testBuild, 2)
+      // Don't pass in any settings so that the previous ones are used instead
+      testBuild.state.build.checkForChange(None, logger).map {
+        case Build.ReturnPreviousState =>
+          sys.error(s"Expected return updated state, got previous state")
+        case action: Build.UpdateState =>
+          assert(action.createdOrModified.size == 2)
+          assert(action.deleted.size == 0)
+          assert(action.invalidated.size == 0)
+          assert(action.settingsForReload == Some(sameSettings))
+      }
+  }
+
+  testLoad("reload if two file contents changed with same settings", Some(sameSettings)) {
```

Comment:
I added a few more tests here.I **really** appreciate that you were the first one to add tons of tests here, I had almost forgotten about this test suite :sweat_smile: